src/rdkafka_cgrp.c

```diff
@@ -2434,6 +2435,280 @@ static void rd_kafka_cgrp_offset_commit_tmr_cb (rd_kafka_timers_t *rkts,
+                (rkcg->rkcg_assignment ? rkcg->rkcg_assignment->cnt : 0);
+
+        rd_assert(partitions);
+
+        /* If the consumer has raised a fatal error, remove the entire
+         * assignment */
+        if (rd_kafka_fatal_error_code(rkcg->rkcg_rk)) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER,
+                             "UNASSIGN", "Group \"%s\": consumer is "
+                             "in a failed state, treating incremental "
+                             "unassign of %d partitions as unassign of all "
+                             "%d partitions in the current assignment",
+                             rkcg->rkcg_group_id->str,
+                             partitions->cnt, !rkcg->rkcg_assignment ?
+                             0 : rkcg->rkcg_assignment->cnt);
+                err = rd_kafka_cgrp_assign(rkcg, NULL);
+                rd_assert(err);
+                return rd_kafka_error_new(err, rd_kafka_err2str(err));
+        }
+
+        if (rd_atomic32_get(&rkcg->rkcg_assignment_lost) &&
+            partitions->cnt < cur_assignment_cnt) {
+                return rd_kafka_error_new(
+                        RD_KAFKA_RESP_ERR__INVALID_ARG,
+                        "Current assignment is lost, but incremental "
+                        "unassign only includes %d of %d partitions",
+                        partitions->cnt, cur_assignment_cnt);
+        }
+
+        /* Verify partitions exist in current assignment before making
+         * any changes to it to ensure the operation doesn't partially
+         * succeed. */
+        for (i = 0; i < partitions->cnt; i++) {
+                const rd_kafka_topic_partition_t *rktpar =
+                        &partitions->elems[i];
+                if (!rd_kafka_topic_partition_list_find(rkcg->rkcg_assignment,
+                                                        rktpar->topic,
+                                                        rktpar->partition)) {
+                        return rd_kafka_error_new(
+                                RD_KAFKA_RESP_ERR__INVALID_ARG,
+                                "%s [%"PRId32"] is not part of the current "
+                                "assignment",
+                                rktpar->topic, rktpar->partition);
+                }
+        }
+
+        /* Whether or not it was before, current assignment is now not lost. */
+        rd_atomic32_set(&rkcg->rkcg_assignment_lost, rd_false);
+
+        if (partitions->cnt == 0) {
+                rd_kafka_dbg(rkcg->rkcg_rk, CGRP|RD_KAFKA_DBG_CONSUMER, "UNASSIGN",
+                     "Group \"%s\": removing 0 of %d assigned partition(s) "
+                     "in join state %s (nothing to do)",
+                     rkcg->rkcg_group_id->str, rkcg->rkcg_assignment->cnt,
+                     rd_kafka_cgrp_join_state_names[rkcg->rkcg_join_state]);
+                return NULL;
+        }
+
+        rd_kafka_cgrp_version_new_barrier(rkcg);
```

Comment:
please consider from here down to the end of the method carefully. I'm not 100% confident I haven't missed something important.